### PR TITLE
Add Parser aspect for match expressions

### LINKED ISSUE #38: Parse match expressions
parse match expression with basic pattern match and case guards
```scala
match $1 {
   -e                  => echo argument is -e
   arg @ -f | -g | -d  => { 
      echo $arg 
      echo b
   }
      //possibility to add guards
   x @ -j | -e if [ 1 == 2 ] || cmd => ...
   *                               => echo $it //fallback with $it being the matched arg
}
```

### PR DESCRIPTION
**Pull request by Override-6**
Fixes #38

### CODE REVIEW

File: parser/src/aspects/match.rs [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+        }
+
+        let first = self.parse_pattern()?;
+
+        if first == Wildcard {
+            return if !is_at_pattern_end!() {
+                self.expected("wildcard pattern cannot be followed by other patterns")
+            } else {
+                Ok(vec![first])
+            }
```
Comment (syldium) [inline parser/src/aspects/match.rs:100]:
```suggestion
            return if is_at_pattern_end!() {
                Ok(vec![first])
            } else {
                self.expected("wildcard pattern cannot be followed by other patterns")
            }
```
Don't make unnecessary condition negations.

File: parser/src/aspects/match.rs [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+        }
+    }
+
+    fn parse_guard(&mut self) -> ParseResult<Option<Expr<'a>>> {
+        if self.cursor.advance(aerated(of_type(If))).is_none() {
+            return Ok(None)
+        }
+
+        self.expression().map(Some)
```
Comment (syldium) [inline parser/src/aspects/match.rs:144]:
```suggestion
        if self.cursor.advance(aerated(of_type(If))).is_none() {
            Ok(None)
        } else {
            self.expression().map(Some)
        }
```

File: parser/src/aspects/match.rs [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+    fn parse_match_arms<P>(&mut self, parse_arm: P) -> ParseResult<Vec<MatchArm<'a>>>
+        where P: FnMut(&mut Self) -> ParseResult<Expr<'a>> + Clone {
+        self.cursor.force(blanks().then(of_type(CurlyLeftBracket)), "expected '{'")?;
+
+        let mut arms: Vec<MatchArm<'a>> = Vec::new();
+
+        while self.cursor.advance(blanks().then(of_type(CurlyRightBracket))).is_none() {
+            let arm = self.parse_match_arm(parse_arm.clone())?;
+            arms.push(arm);
+        }
```
Comment (syldium) [inline parser/src/aspects/match.rs:47]:
```suggestion
        while self.cursor.lookahead(blanks().then(eod())).is_none() {
            let arm = self.parse_match_arm(parse_arm.clone())?;
            arms.push(arm);
        }
        self.cursor.force(blanks().then(of_type(CurlyRightBracket)), "expected '}'")?;
```
The parser should stop at each closing punctuation to report unpaired delimiters (see #33).

**Review by syldium — CHANGES_REQUESTED** [review]
The parser is confused with this input and does not stop after an invalid operand:
```rust
match nginx {
    * => echo $it
}
```
Message: `Unexpected token '=>'.`

**Review by syldium — APPROVED** [review]
LGTM

File: parser/src/aspects/literal.rs [hunk 4]
```diff
@@ -71,34 +90,33 @@ impl<'a> LiteralAspect<'a> for Parser<'a> {
...
         let mut lexeme = self.cursor.peek().value;
         let mut literal_value = String::new();
         let mut parts = Vec::new();
         loop {
             if self.cursor.is_at_end() {
                 return self.expected(
                     "Unterminated string literal.",
-                    ParseErrorKind::Unpaired(self.cursor.relative_pos(&start)),
+                    ParseErrorKind::Unpaired(self.cursor.relative_pos(lexeme)),
                 );
```
Comment (syldium) [inline parser/src/aspects/literal.rs:110]: The unpaired error should point at the unpaired token, i.e. the quote. If the lexeme is used, the start annotation is wrong.
```
  × Unterminated string literal.
   ╭─[stdin:1:1]
 1 │ "aaa
   ·  ─┬─┬
   ·   │ ╰── Here
   ·   ╰── Start
   ╰────
```
The `Unexcepted` error would be more appropriate if you want to annotate the whole lexeme:
```rs
return self.expected_with(
    "Unterminated string literal.",
    lexeme,
    ParseErrorKind::Unexpected
);
```
This also applies to the `string_literal` method.

File: parser/src/aspects/literal.rs [hunk 5]
```diff
@@ -2,49 +2,68 @@ use context::source::try_join_str;
...
 use crate::aspects::substitution::SubstitutionAspect;
-use lexer::token::TokenType;
+use lexer::token::TokenType::*;
 
-use crate::ast::literal::{Literal, LiteralValue};
+use crate::ast::value::{Literal, LiteralValue, TemplateString};
 use crate::ast::*;
 use crate::err::ParseErrorKind;
-use crate::moves::{next, of_type, word_sep};
+use crate::err::ParseErrorKind::{Unexpected};
```
Comment (syldium) [inline parser/src/aspects/literal.rs:10]: I don't think that importing the enum variant directly is really needed here.

File: parser/src/aspects/match.rs [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+    }
+}
+
+impl<'a> Parser<'a> {
+    fn parse_match_arms<P>(&mut self, parse_arm: P) -> ParseResult<Vec<MatchArm<'a>>>
+    where
+        P: FnMut(&mut Self) -> ParseResult<Expr<'a>> + Clone,
+    {
+        self.cursor
+            .force(blanks().then(of_type(CurlyLeftBracket)), "expected '{'")?;
```
Comment (syldium) [inline parser/src/aspects/match.rs:48]:
```suggestion
        let opening_bracket = self.cursor.force_with(
            blanks().then(of_type(CurlyLeftBracket)),
            "expected match start",
            ParseErrorKind::Excepted("{"),
        )?;
        self.delimiter_stack.push_back(opening_bracket.clone());
```
This helps the parser tracking delimiter pairs and reporting better errors.

File: parser/src/aspects/match.rs [hunk 7]
```diff
@@ -0,0 +1,565 @@
...
+
+        let mut arms: Vec<MatchArm<'a>> = Vec::new();
+
+        while self.cursor.lookahead(blanks().then(eod())).is_none() {
+            let arm = self.parse_match_arm(parse_arm.clone())?;
+            arms.push(arm);
+        }
+        self.cursor
+            .force(blanks().then(of_type(CurlyRightBracket)), "expected '}'")?;
```
Comment (syldium) [inline parser/src/aspects/match.rs:57]:
```suggestion
        self.cursor.force_with(
            blanks().then(of_type(CurlyRightBracket)),
            "expected '}'",
            ParseErrorKind::Unpaired(self.cursor.relative_pos(opening_bracket)),
        )?;
        self.delimiter_stack.pop_back();
```

**Review by syldium — CHANGES_REQUESTED** [review]

File: parser/src/aspects/literal.rs (hunk 4 above)
Comment (Override-6) [reply parser/src/aspects/literal.rs:110]: prefer it with the 'start' cursor, i'll fix that

File: parser/src/moves.rs [hunk 8]
```diff
@@ -153,17 +154,36 @@ pub(crate) fn spaces() -> RepeatedMove<PredicateMove<impl (for<'a> Fn(Token<'a>)
...
+        PredicateMove<impl (for<'a> Fn(Token<'a>) -> bool) + Copy>,
         PredicateMove<impl (for<'a> Fn(Token<'a>) -> bool) + Copy>,
-        AndThenMove<
-            PredicateMove<impl (for<'a> Fn(Token<'a>) -> bool) + Copy>,
-            PredicateMove<impl (for<'a> Fn(Token<'a>) -> bool) + Copy>,
-        >,
     >,
 > {
-    repeat_n(1, space().or(of_type(BackSlash).and_then(of_type(NewLine))))
+    space().or(of_type(BackSlash).and_then(of_type(NewLine)))
```
Comment (syldium) [inline parser/src/moves.rs:165]: Is this change intentional? You are literally using `repeat_n(1,word_sep())` in the literal parser.

File: parser/src/ast/value.rs [hunk 9]
```diff
@@ -15,6 +16,11 @@ pub enum LiteralValue {
     Float(f64),
 }
 
+#[derive(Debug, Clone, PartialEq, DebugPls)]
+pub struct TemplateString<'a> {
```
Comment (syldium) [inline parser/src/ast/value.rs:20]:
```suggestion
/// A group of expressions that can be interpolated into a string.
#[derive(Debug, Clone, PartialEq, DebugPls)]
pub struct TemplateString<'a> {
```

File: parser/src/aspects/if_else.rs [hunk 10]
```diff
@@ -92,6 +92,7 @@ mod tests {
 
     #[test]
     fn if_else_if() {
+
```
Comment (syldium) [inline parser/src/aspects/if_else.rs:95]:
```suggestion
```

File: parser/src/aspects/if_else.rs [hunk 11]
```diff
@@ -130,6 +131,7 @@ mod tests {
 
     #[test]
     fn if_else_if_separations() {
+
```
Comment (syldium) [inline parser/src/aspects/if_else.rs:134]:
```suggestion
```

File: parser/src/aspects/if_else.rs [hunk 12]
```diff
@@ -172,6 +174,7 @@ mod tests {
 
     #[test]
     fn if_else_as_value() {
+
```
Comment (syldium) [inline parser/src/aspects/if_else.rs:177]:
```suggestion
```

File: parser/src/aspects/if_else.rs [hunk 13]
```diff
@@ -201,27 +204,29 @@ mod tests {
...
+                    }))),
+
                 }))),
             }),]
         )
     }
 
     #[test]
     fn if_else_bad_brackets() {
+
```
Comment (syldium) [inline parser/src/aspects/if_else.rs:221]:
```suggestion
```

**Review by syldium — CHANGES_REQUESTED** [review]

**Review by syldium — APPROVED** [review]
